### PR TITLE
c-api: add wasi_config_set_stdout_bytes

### PR DESCRIPTION
**Pull request by ShuP1**
As discussed in issue #4372 and respective issues about python and go bindings.

Can replace `wasi_config_inherit_stdout` in [examples/wasi/main.c:79](https://github.com/bytecodealliance/wasmtime/blob/e6d339b6ac343326c86d1d193cee6e2db96fa0bc/examples/wasi/main.c#L79)
```c
wasi_config_t *wasi_config = wasi_config_new();
assert(wasi_config);
wasi_out_bytes_t *wasi_stderr = wasi_config_set_stderr_bytes(wasi_config, 32);
error = wasmtime_context_set_wasi(context, wasi_config);
// run and delete store
wasm_byte_vec_t wasi_message;
wasi_out_bytes_take(wasi_stderr, &wasi_message);
printf("stderr:\n%.*s\n", (int)wasi_message.size, wasi_message.data);
```
See https://github.com/bytecodealliance/wasmtime/commit/3174d9a9928b756c3427403096a05c87bb2bb988

(cc @alexcrichton)

### CODE REVIEW

**Review by alexcrichton — COMMENTED** [review]
Thanks! API-wise though I've got a few thoughts. Would it make more sense to have the operation be something along the lines of "please capture stdout/stderr" and then later you can sort of "drain" it to get the output?

Specifying the size up-front seems somewhat unfortunate here since specifying a small size limits what the instance can print, but specifying a large size is wasteful if not much is printed. Instead if the writer specified is simply `Vec<u8>` I think that could be avoided?

Furthermore for acquiring the output one option would be to have a buffer passed in for a `read`-style API intead of a "give me your buffer" sytle API. That way applications could incrementally read bytes. Alternatively there could be an advancing-style API where you ask "give me your internal pointer and size" and later say "ok please advance by N bytes". 

Would such an API work for your use case?

**Comment by ShuP1** [comment]
Good points. I was planning to add a pipe in a later PR.
Basically a `VecDeque<u8>` with read/write interface both for Wasi and API.

In my use case, the "read" API is perfect and easy to implement since `wasi_common::pipe` are `Arc<RwLock<T>>` anyway.
It's not a "superset" since Vec is unbounded. A program allocating GBs by flooding stdout is problematic IMO.

Still WIP: https://github.com/bytecodealliance/wasmtime/compare/main...ShuP1:wasmtime:c-api-stdio-pipe-example

In conclusion, this PR can probably be replaced by the next proposal with a capacity limit to the queue.

**Comment by alexcrichton** [comment]
That also seems plausible to me yeah, and the high-output case could possibly be handled with an optional maximum on the pipe buffer?

I think it would be good to document some more the semantics of the pipes, for example:

* Are the pipes safe to span threads?
* What happens when a write hits the capacity threshold?

I might also recommend a few changes such as:

* I'd avoid `wasm_byte_vec_t` as it's cumbersome to work with and instead use `read`/`write`-style buffer passing
* I'd recommend splitting the pipe into halves and taking ownership when inserting it into a `wasi_config_t` to avoid the same write half being used in two different instances.

**Comment by ShuP1** [comment]
> Are the pipes safe to span threads?
Safe to move to another thread but may lock.

Or even maybe dead lock during concurrent use ?
This reaches the limit of my understand of rust async...

https://docs.rs/wasi-common/latest/src/wasi_common/pipe.rs.html#112 
`RwLock` is locked in an async function who does **not** use await.
So no "context switch" ? and no risk for another thread to hung on this lock ?

**Comment by ShuP1** [comment]
> I'd avoid wasm_byte_vec_t as it's cumbersome to work with and instead use read/write-style buffer passing

Does this point also includes changing that ?
```c
void wasi_config_set_stdin_bytes(wasi_config_t* config, wasm_byte_vec_t* binary);
// to
void wasi_config_set_stdin_bytes(wasi_config_t* config, size_t, const char*);
```
